docs/client-design.md

```diff
@@ -0,0 +1,306 @@
+	 * Return the connection's state.
+	 */
+	enum ConnectionState get_state() const;
+	
+	/** 
+	 * If the connection is an erroneous state (see `status`), return the 
+	 * connection error that caused it. The same error is also passed to 
+	 * response callbacks, and the same error is returned by 
+	 * `Request::get_error`. 
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/* An abstract request's interface. A request object is always returned. */
+	Request some_request(/* options */);
+};
+```
+
+#### Connection State
+
+```c++
+enum ConnectionState {
+	CONNECTION_INITIAL = 0,
+	CONNECTION_AUTH = 1,
+	CONNECTION_ACTIVE = 2,
+	CONNECTION_ERROR = 3,
+	CONNECTION_ERROR_RECONNECT = 4,
+};
+```
+
+#### Connect Options
+
+```c++
+/* Extend `ConnectionOptions` with an option for the reconnection feature. */
+struct ConnectOptions {
+	/* All existing options. */
+	
+	/** 
+	 * In the event of a broken connection, the interval in which the stream 
+	 * tries to re-establish the connection. 
+	 */
+	static constexpr size_t DEFAULT_RECONNECTION_INTERVAL = 2;
+	size_t reconnection_interval = DEFAULT_RECONNECTION_INTERVAL;
+};
+```
+
+### Requests
+
+```c++
+/** Encapsulates management of a request issued through a connection. */
+class Request {
+public:
+	/** 
+	 * Callback called when response is ready. Since the callback has a fixed 
+	 * signature, we need to allow for capturing additional context using 
+	 * lambdas. Hence, we use `std::function` for type erasure.
+	 * 
+	 * See `Connection::get_error` for details about the `error` parameter.
+	 */ 
+	using request_cb_f = 
+		std::function<void(Request &&request, Connection &connection)>;
+	
+	/** Set a callback called when the response is ready. */
+	void set_callback(request_cb_f request_cb) &&;
+	
+	/** Return the request's status. */
+	enum RequestStatus get_status();
+	
+	/**
+	 * Return a connection error, if any. See `Connection::get_error` for 
+	 * details.
+	 */
+	std::optional<ConnectionError> &get_error() & const;
+	
+	/**
+	 * Cancel the request, ending the lifetime of the corresponding response.
+	 */
+	void cancel();
+	
+	/** 
+	 * Return the response, if any. A response is available, iff:
+	 * 1. The response has been received by the connection.
+	 * 2. The response has not been dispatched to a callback.
+	 * 3. The response has not already been retrieved previously.
+	 */
+	std::optional<Response> get_response();
+};
+```
+
+#### Request Status
+
+```c++
+enum RequestStatus {
+	REQUEST_SUCCESS = 0,
+	REQUEST_ERROR = -1,
+	REQUEST_IN_PROGRESS = 1,
+};
+```
+
+#### Request Fan-Out
+
+```c++
+/**
+ * A fan-out encapsulates the collection of responses for a collection of 
+ * requests.
+ */
+class FanOut {
+	template<class InputIt>
+	FanOut(InputIt first, InputIt last)
+	FanOut(std::initializer_list<Request> requests);
+	
+	/**
+	 * Return a list of ready request, and cancel the requests for which a 
+	 * response is not available. 
+	 */
+	tnt::List<Request> collect();
```

Comment:
Is this solution of myTarget's problem? If so:

1. You mustn't cancel requests here.
2. Passed requests must be dynamic: you poll for ready requests, then you create a bunch of new requests and poll them along with the previous ones.
3. From the previous point - you must allow to populate the list of requests.
4. Need to come up with a better name - it is a tool polling for ready request, not for "FanOut" requests.

P.S. The idea was to create something like `epoll` - you ask for ready requests and get only them (no need to scan for all requests and find ready ones), but all not ready requests must be still available.

Reply:
Perhaps I still don't understand the problem statement. AFAIC, there is no dynamic scenario, we just want to create a bunch of synchronization points (i.e., `collect` points). 

In the scenario you propose the mapping between the business logic and the responses is not clear. Let's say I have requests about a VK and Mail.ru profile mixed up. How I am going to understand what the sources of the requests were?

Reply:
I'll explain my understanding from the beginning.

We have myTarget's problem - collect as much information as we can in a limited amount of time, for example, 1 second. We have only anonymous user id, so, firstly we send a bunch of requests to Tarantool to get all actions with this id (email authorization, visiting book page on litres and so on). When we have received data from `mail.ru` email, we can send a new bunch of requests to Tarantool storing `mail.ru` data (name, surname, age, etc). When we have received data from litres with user's books ids, we can send another bunch to Tarantool storing litres data (we want to find book's genre by its id). After 1 second all the requests are canceled, and myTarget knows that the guy with `session_id = 143` is actually Ivan Ivanov who is keen on Steven King books.

Firstly, I just can't implement such service with this `FanOut`. For example, I call `collect` after 0.1s, and no responses are ready - the job has died even before it started. So, we **must not** cancel requests that are not ready.

Secondly, I have many different request types (search by anonymous id, `mail.ru`, `litres` and so on). I see only two solutions:
1. We will have a lot of `FanOut`s - one per request type for every job (collecting info by id): `N*M` combinators.
2. We will have something like `epoll` - the combinator will associate a value with every request. It can be `void *` or a template parameter of the combinator. Then, every job will wake up every 0.05s, for example, and call single `collect` method for all request types. Here we will use only `N` combinators.

I would prefer the second variant.

Thirdly, let's imagine that we have access to Tarantool storing Delivery Club order history. Now myTarget knows my user id in Delivery Club and wants to fetch all my orders. By user id it fetches a lot of reponses with order id. Then, by order id, it fetches information about food to find out about my food preferences. If we decide to have a `FanOut` per request type, we will need an opportunity to append new requests resolving `order_id` into its contents - collection of requests of this type is filled dynamically.